### ISSUE TITLE
Failed MD5 integrity check loading fonts

### ISSUE THREAD

**Issue by JoaoBGusmao** [expo/expo#5566]
### Environment

 Expo CLI 3.0.10 environment info:
    System:
      OS: macOS 10.14.6
      Shell: 5.3 - /bin/zsh
    Binaries:
      Node: 12.8.1 - /usr/local/bin/node
      Yarn: 1.15.2 - ~/.yarn/bin/yarn
      npm: 6.10.2 - /usr/local/bin/npm
      Watchman: 4.7.0 - /usr/local/bin/watchman
    IDEs:
      Android Studio: 3.5 AI-191.8026.42.35.5791312
      Xcode: 10.2/10E125 - /usr/bin/xcodebuild
    npmGlobalPackages:
      expo-cli: 3.0.10

Target application: **android**
### Steps to Reproduce
- Clone the repo located at reproducible demo section bellow.
- Install dependencies
- Get an real android device
- Clear expo client cache
- run expo start on packages/app

The bug should occur right after the app finish loading the bundle

+ info about the error and how we bypassed it on demo readme

### Expected Behavior
Should load the required font even if it is located inside another package

### Actual Behavior
failed MD5 integrity check only when the font we are requiring is located outside our app node_modules (a common case when using monorepo)

### Reproducible Demo
Here is a reproducible demo using monorepo.
https://github.com/JoaoBGusmao/md5-bug

**Comment by cruzach** [comment]
Hey @JoaoBGusmao sorry you're running into this

Can you take a look at [this issue](https://github.com/expo/expo/issues/4626), they had a similar issue/fix with the passing explicit paths to get into node_modules manually

**Comment by JoaoBGusmao** [comment]
@cruzach I can confirm that passing explicit path works, but in our use case, it wont.
It wont work only when the font is pointing to a node_modules of a package different from the one that is running.

I opened this issue because this is something that was working fine before 33, and this keep our FontLoader reusable. We were able to fix it by duplicating the imports in all packages that use FontLoader
```
<FontLoader
    fonts={{
        Roboto: require('../node_modules/@joaobgusmao/typeface-roboto/files/Roboto-Bold.ttf'),
        Roboto_bold: require('../node_modules/@joaobgusmao/typeface-roboto/files/Roboto-Bold.ttf'),
        Roboto_medium: require('../node_modules/@joaobgusmao/typeface-roboto/files/Roboto-Medium.ttf'),
        streamline_icons: require('../node_modules/@joaobgusmao/streamline-icons/fonts/streamline_icons.ttf'),
    }}
>
    ...
```
But this sadly breaks the reusable purpose of FontLoader

By the way, I believe this issue will be fixed when #4626 is fixed

**Comment by DanielTate** [comment]
FYI I fixed this issue in my project by changing the font names to remove spaces...
No idea what's going on here...

Before FAILING:
````
    'proxima-nova-black': require('./assets/fonts/Proxima Nova Black.ttf'),
````

This failed exactly as you described above. Although it was random and I couldn't in it down.
After about an hour of messing around I just changed the name of the font and it works 100% of the time now... for now anyway.

After WORKING:
````
    'proxima-nova-black': require('./assets/fonts/proxima-nova-black.ttf'),

**Comment by cruzach** [comment]
Yeah, that's actually not to rare of a problem for folks. We should update the docs to say that font filenames should not include any spaces

**Comment by JoaoBGusmao** [comment]
@DanielTate @cruzach this issue is related to the use of monorepo. It will happen with any filename. The reproducible demo here uses monorepo and any filename will fail with the same error.

It wasn't the ideal approach for my project, but requiring our fonts on main package and passing it to the package that uses the expo function fixed this issue.

**Comment by ShaggyKris** [comment]
```
/ Load any resources or data that we need prior to rendering the app
    React.useEffect(() => {
        async function loadResourcesAndDataAsync() {
            try {
                SplashScreen.preventAutoHide();

                // Load our initial navigation state
                setInitialNavigationState(await getInitialState());

                // Load fonts
                await Font.loadAsync({
                    ...Ionicons.font,
                    ...MaterialIcons.font,
                    ...MaterialCommunityIcons.font,
                    "space-mono": require("./assets/fonts/SpaceMono-Regular.ttf"),
                });
                await registerForPushNotificationsAsync();
            } catch (e) {
                // We might want to provide this error information to an error reporting service
                console.warn(e);
            } finally {
                setLoadingComplete(true);
                SplashScreen.hide();
            }
        }

        loadResourcesAndDataAsync();
    }, []);
```

This fails when called at app root `React.useEffect()` using the Expo Tabs default project. I get the following error:

```
[Unhandled promise rejection: Error: Downloaded file for asset 'MaterialCommunityIcons.ttf' Located at http://[redacted IP]:19001/assets/node_modules/@expo/vector-icons/build/vendor/react-native-vector-icons/Fonts/MaterialCommunityIcons.ttf?platform=android&hash=5a293a273bee8d740a045d9922b9a9ae?platform=android&dev=true&minify=false&hot=false failed MD5 integrity check]
- node_modules/expo-asset/build/PlatformUtils.js:49:28 in _downloadAsyncManagedEnv
- node_modules/regenerator-runtime/runtime.js:45:44 in tryCatch
- node_modules/regenerator-runtime/runtime.js:274:30 in invoke
- node_modules/regenerator-runtime/runtime.js:45:44 in tryCatch
- node_modules/regenerator-runtime/runtime.js:135:28 in invoke
- node_modules/regenerator-runtime/runtime.js:145:19 in PromiseImpl.resolve.then$argument_0
- node_modules/promise/setimmediate/core.js:37:14 in tryCallOne
- node_modules/promise/setimmediate/core.js:123:25 in setImmediate$argument_0
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:146:14 in _callTimer
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:194:17 in _callImmediatesPass
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:458:30 in callImmediates
* [native code]:null in callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:407:6 in __callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:143:6 in __guard$argument_0
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:384:10 in __guard
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:142:17 in __guard$argument_0
* [native code]:null in flushedQueue
* [native code]:null in invokeCallbackAndReturnFlushedQueue
```

**Comment by brentvatne** [comment]
@ShaggyKris - can you post your project exactly as-is for us to try to repro that?

**Comment by tonnenpinguin** [comment]
I just updated expo sdk to the latest version (37.0.11) and started running into the same issue.

The reason seems to be that the double dots are not handled correctly (opening the url in chrome actually redirects to `http://192.168.0.18:19001/node_modules/@expo/vector-icons/build/vendor/react-native-vector-icons/Fonts/MaterialCommunityIcons.ttf?platform=android&hash=3c851d60ad5ef3f2fe43ebd263490d78?platform=android&dev=true&minify=false&hot=false` stripping the double dots that navigated to the root node_modules of the umbrella app.

Adding the @expo/vector-icons package to lernas nohoist list seems to do the trick for us.

**Comment by bernardnongpoh** [comment]
I found the solution by loading the font manually. 
Download the font from [MaterialCommunityIcons.ttf](https://github.com/oblador/react-native-vector-icons/blob/master/Fonts/MaterialCommunityIcons.ttf) and place it in the assets folder (`./assets/fonts/MaterialCommunityIcons.ttf`). At the root of the application, load the font. It works.
```

import * as Font from 'expo-font';
import { Main } from './src/main';
import { loadAsync } from 'expo-font';

export default function App() {
  useEffect(() => {

    async function loadFont() {
    await loadAsync({
      // Load a font `Montserrat` from a static resource
      'material-community': require('./assets/fonts/MaterialCommunityIcons.ttf'),
    
     
    });
  
  }
  loadFont();
   
  }, []);

 

  return (
    <SafeAreaProvider>
      <AppearanceProvider>
        <Main />
      </AppearanceProvider>
    </SafeAreaProvider>
  );
}
```

**Comment by bernardnongpoh** [comment]
One more solution you can try without using `useEffect()`. 

```
import React, { useEffect, useState } from 'react';
import * as Font from 'expo-font';
import { Main } from './src/main';
import { loadAsync } from 'expo-font';
import {AppLoading} from 'expo';
export default function App() {

  const [loadedFont,setLoadedFont]= useState(false);
  async function loadFont() {
    await loadAsync({
      // Load a font `Montserrat` from a static resource
      'material-community': require('./assets/fonts/MaterialCommunityIcons.ttf'),
    
     
    });
  
  }

    if(!loadedFont)
      return<AppLoading
        startAsync={loadFont}
        onFinish={()=>setLoadedFont(true)}
     />
    return (
    
       <Main />
    );
}
```

**Comment by anasanzari** [comment]
any update on this?

**Comment by ShaggyKris** [comment]
@anasanzari 

So my issue seemingly went away, but I'm not sure how. I believe it was due to a mismatched version of @expo/vector-icons, but I can't be sure.

**Comment by anasanzari** [comment]
@ShaggyKris I have fixed. Actually the problem was with the bundler. I have a monorepo setup with yarn workspaces, where fonts were loaded from a sub package. Bundler has problems with loading assets from subpackages.

**Comment by vpicone** [comment]
@anasanzari I'm using the same setup, any insights on how to resolve this?

**Comment by JoaoBGusmao** [comment]
@vpicone as I said before, you must require all your fonts in your main package (I mean, require as the package expo is installed). This should work.

This bug happens when a sub package tries to require the font file

**Comment by vpicone** [comment]
@JoaoBGusmao what about for a component library that wants to use a specific font? Should all the app developers need to require that font? And even then only to accommodate Android weirdness?

**Comment by JoaoBGusmao** [comment]
@vpicone usually, libs that use custom fonts doesn't import them in their own code. They ask you yo do it through Font.loadAsync on your App.js. This bug would only occur in cases that Font.loadAsync is called by another package, and since libs doesn't use it, I believe this won't be a problem

**Comment by vpicone** [comment]
@JoaoBGusmao thanks for the context, I was hoping to abstract the font management away from devs but it looks like that’s not the way things are typically done.

**Comment by anasanzari** [comment]
@vpicone sorry, didn't catch the notification. I actually did a hack to fix the imports. Bundler doesn't serve any files outside the folder `expo start` command is executed and files you're trying to import actually reside outside.
If you're using lerna, you can add the subpackage to `nohoist` in your package.json and files can be successfully imported in your package.

**Comment by HelixHEX** [comment]
Hi im currently using expo inside turborepro (monorepo) and I've run into the same problem except I'm not using custom icons from @expo/vector-icon just default ones. I've attached how I'm importing the component.

<img width="493" alt="Screen Shot 2021-12-18 at 12 05 26 AM" src="https://user-images.githubusercontent.com/25518600/146629816-5206daf0-01fc-45d1-bb43-63efd368c10c.png">
s

**Comment by MusfiquerRhman** [comment]
> FYI I fixed this issue in my project by changing the font names to remove spaces... No idea what's going on here...
> 
> Before FAILING:
> 
> ```
>     'proxima-nova-black': require('./assets/fonts/Proxima Nova Black.ttf'),
> ```
> 
> This failed exactly as you described above. Although it was random and I couldn't in it down. After about an hour of messing around I just changed the name of the font and it works 100% of the time now... for now anyway.
> 
> After WORKING:
> 
> ```
>     'proxima-nova-black': require('./assets/fonts/proxima-nova-black.ttf'),
> ```

I had a comma in the file name, and removing it solved the issue.

**Comment by tomrom-pomelo** [comment]
This is still consistently happening for me for all assets on *Android* only. If the asset is in a shared location, whether it be a font from a dependency or an image from a folder outside the project directory, it will return a `failed md5 integrity check` error. I posted an issue to https://github.com/expo/google-fonts/issues/75 , but after seeing that this also affects local images in shared (non-project) directories in our monorepo, I think it's a larger problem.

Of course I could pass down images to the non-project directory as others suggested, but it's not ideal given that components could be many nested levels deep. My only other idea is to create an `AssetContext` that shared components can get images from, and the provider is defined at the project level. 

We are not using lerna so I can't use @anasanzari's suggestion.

Anyone have any other ideas here? Or should I re-open this issue since it still affects projects?

**Comment by juliushuck** [comment]
@tomrom-pomelo I also have the problem on Android but works on iOS. I'm using a turborepo. For me, the fix by @bernardnongpoh (https://github.com/expo/expo/issues/5566#issuecomment-631228248) worked. But there is no way this is the intended behavior, so I would also like to see this to get fixed.

**Comment by Wolfleader101** [comment]
Unsure why this is closed - it is still an issue.

I have the same issue when using a turborepo.

**Comment by igoro00** [comment]
@Wolfleader101 @HelixHEX and others
Okay, I fixed it. Expo expects to pull the fonts from the **local** node_modules. So we need to download @expo/vector-icons in that node_modules folder. 
Here's the SO page I found that says how to do this: https://stackoverflow.com/questions/56910467/when-using-yarn-workspaces-how-to-force-a-package-to-be-installed-in-the-relati

### TLDR
Put this in your local `package.json`:
```
"workspaces": {
  "nohoist": [
    "@expo/vector-icons"
  ]
}
```
and run `yarn install`

**Comment by fragilebanana16** [comment]
my issue is to play mp3 file, some are just fine, but others not. However, when i delete the space inside file name, it worked, but its still not quite elegant to do so, any better ideas?

**Comment by maersk96** [comment]
I am experiencing same problem - **only on local server and only on Android**.

In the error there is a personal file path to an Expo developer named [Kudo](https://github.com/Kudo).

I don't think this package has ever been ready for production.

![image](https://github.com/expo/expo/assets/7141352/7a27662a-496a-4455-88e8-2d632d9f620f)

**Comment by ste7en** [comment]
I'm still getting this error in a monorepo.
```
[Error: Downloaded file for asset 'Lato-Black.ttf' Located at 
http://127.0.0.1:8081/assets/../../libs/shared/configs/src/fonts/Lato-Black.ttf?
platform=android&hash=e631d2735799aa943d93d301abf423d2?platform=android&dev=true&hot=false failed MD5 integrity check]
```

**Comment by tichopad** [comment]
Howdy!

I've encountered this issue (failed MD5 integrity check) with font files bundled in `@expo/vector-icons`. Specifically FontAwesome.ttf, MaterialIcons.ttf and MaterialCommunityIcons.ttf.

My setup includes a workspace that is an Expo mobile app and another that's purely UI components. Both workspaces had `@expo/vector-icons` as a dependency and I'm using Yarn with node-modules linker. Because of that `@expo/vector-icons` got installed to the root node_modules directory and that might've caused the issue.

I fixed it by adding extra node_modules paths to metro.config.js:
```javascript
const { getDefaultConfig } = require('expo/metro-config');
const path = require('path');

const projectRoot = __dirname;
// My app's workspace is two levels deep (e.g. ./packages/mobile-app)
const workspaceRoot = path.resolve(projectRoot, '../..');

const defaultConfig = getDefaultConfig(projectRoot);

// Enable Metro to resolve packages to both local and workspace node_modules in this order
defaultConfig.resolver.nodeModulesPaths = [
  path.resolve(projectRoot, 'node_modules'),
  path.resolve(workspaceRoot, 'node_modules'),
];

module.exports = defaultConfig
```

Btw check [this repository](https://github.com/byCedric/expo-monorepo-example/tree/main) for an example of a proper monorepo setup with Expo. Props to the wonderboy @byCedric for saving my bacon once again.

**Comment by nilooy** [comment]
> @Wolfleader101 @HelixHEX and others Okay, I fixed it. Expo expects to pull the fonts from the **local** node_modules. So we need to download @expo/vector-icons in that node_modules folder. Here's the SO page I found that says how to do this: https://stackoverflow.com/questions/56910467/when-using-yarn-workspaces-how-to-force-a-package-to-be-installed-in-the-relati
> 
> ### TLDR
> Put this in your local `package.json`:
> 
> ```
> "workspaces": {
>   "nohoist": [
>     "@expo/vector-icons"
>   ]
> }
> ```
> 
> and run `yarn install`

this worked for me
